src/nvim/os/env.c

```diff
@@ -87,3 +97,598 @@ void os_get_hostname(char *hostname, size_t len)
+
+  if (!one)
+    src = skipwhite(src);
+  while (*src && dstlen > 0) {
+    /*
+     * Here we are at the beginning of a file name.
+     * First, check to see if the beginning of the file name matches
+     * $HOME or the "real" home directory. Check that there is a '/'
+     * after the match (so that if e.g. the file is "/home/pieter/bla",
+     * and the home directory is "/home/piet", the file does not end up
+     * as "~er/bla" (which would seem to indicate the file "bla" in user
+     * er's home directory)).
+     */
+    p = homedir;
+    len = dirlen;
+    for (;; ) {
+      if (   len
+             && fnamencmp(src, p, len) == 0
+             && (vim_ispathsep(src[len])
+                 || (!one && (src[len] == ',' || src[len] == ' '))
+                 || src[len] == NUL)) {
+        src += len;
+        if (--dstlen > 0)
+          *dst++ = '~';
+
+        /*
+         * If it's just the home directory, add  "/".
+         */
+        if (!vim_ispathsep(src[0]) && --dstlen > 0)
+          *dst++ = '/';
+        break;
+      }
+      if (p == homedir_env)
+        break;
+      p = homedir_env;
+      len = envlen;
+    }
+
+    /* if (!one) skip to separator: space or comma */
+    while (*src && (one || (*src != ',' && *src != ' ')) && --dstlen > 0)
+      *dst++ = *src++;
+    /* skip separator */
+    while ((*src == ' ' || *src == ',') && --dstlen > 0)
+      *dst++ = *src++;
+  }
+  /* if (dstlen == 0) out of space, what to do??? */
+
+  *dst = NUL;
+
+  if (homedir_env != homedir_env_orig)
+    free(homedir_env);
+}
+
+/// Like home_replace, store the replaced string in allocated memory.
+/// @param buf When not NULL, check for help files
+/// @param src Input file name
+char_u * home_replace_save(buf_T *buf, char_u *src) FUNC_ATTR_NONNULL_RET
+{
+  size_t len = 3;                      /* space for "~/" and trailing NUL */
+  if (src != NULL)              /* just in case */
+    len += STRLEN(src);
+  char_u *dst = xmalloc(len);
+  home_replace(buf, src, dst, (int)len, true);
+  return dst;
+}
+
+/// Our portable version of setenv.
+/// Has special handling for $VIMRUNTIME to keep the localization machinery
+/// sane.
+void vim_setenv(char_u *name, char_u *val)
+{
+  os_setenv((char *)name, (char *)val, 1);
+  /*
+   * When setting $VIMRUNTIME adjust the directory to find message
+   * translations to $VIMRUNTIME/lang.
+   */
+  if (*val != NUL && STRICMP(name, "VIMRUNTIME") == 0) {
+    char_u  *buf = concat_str(val, (char_u *)"/lang");
+    bindtextdomain(VIMPACKAGE, (char *)buf);
+    free(buf);
+  }
+}
+
+
+/// Function given to ExpandGeneric() to obtain an environment variable name.
+char_u *get_env_name(expand_T *xp, int idx)
+{
+# define ENVNAMELEN 100
+  // this static buffer is needed to avoid a memory leak in ExpandGeneric
+  static char_u name[ENVNAMELEN];
+  // since it's nonsensical for idx to be negative, we will
+  // assume that it is positive and that a cast to size_t is safe
```

Comment:
here too... either `assert()` or don't bother, and in either case leave out the speculative comment.